src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php

```diff
@@ -331,14 +334,22 @@ public function denormalize($data, string $type, string $format = null, array $c
                 }
             }
 
-            $value = $this->validateAndDenormalize($resolvedClass, $attribute, $value, $format, $context);
             try {
-                $this->setAttributeValue($object, $attribute, $value, $format, $context);
-            } catch (InvalidArgumentException $e) {
-                throw new NotNormalizableValueException(sprintf('Failed to denormalize attribute "%s" value for class "%s": '.$e->getMessage(), $attribute, $type), $e->getCode(), $e);
+                $denormalizedValue = $this->validateAndDenormalize($resolvedClass, $attribute, $value, $format, $context);
+                try {
+                    $this->setAttributeValue($object, $attribute, $denormalizedValue, $format, $context);
+                } catch (InvalidArgumentException $e) {
+                    throw new NotNormalizableValueException(sprintf('Failed to denormalize attribute "%s" value for class "%s": '.$e->getMessage(), $attribute, $type), $e->getCode(), $e);
+                }
+            } catch (InvariantViolationException $exception) {
```

Comment:
So, the idea is that all denormalizers need to be aware of the `COLLECT_INVARIANT_VIOLATIONS` context key and behave differently based on this? Either throwing a `NotNormalizableValueException` or a `InvariantViolationException`? Is that correct?

Could we just catch the existing `NotNormalizableValueException`? Or is the problem that these error messages are currently a bit "internal" (e.g. `Failed to denormalize attribute foo for class Foo\Bar`) and so the new exception message allows denormalizers to advertise a safer, "validation" message? Or is there another purpose?

If the only purpose is to advertise a validation-safe message, perhaps we could add a new optional property+method (thinking out loud) to `NotNormalizableValueException` that contains this - e.g. `getValidationMessage()`? Or maybe I'm way off understanding the mechanics :)

Reply:
> So, the idea is that all denormalizers need to be aware of the `COLLECT_INVARIANT_VIOLATIONS` context key and behave differently based on this? Either throwing a `NotNormalizableValueException` or a `InvariantViolationException`? Is that correct?

Yes, I think collecting denormalization errors should be opt-in. Though throwing `InvariantViolationException` is the only implemenation I managed to get working for now.

My first attempt was to wrap all denormalizations using an interface (e.g. `InvariantViolationCatcherInterface`). One implementation would just throw immediatly (current behavior) and another would collect errors instead. But I faced too many issues and gave up with this approach. Maybe I should give it another try?

> Could we just catch the existing `NotNormalizableValueException`?

Here we catch `InvariantViolationException` because we rely on nested denormalizers calls to denormalize nested values. Those denormalizers will only throw this kind of exception for each invariant violation if `COLLECT_INVARIANT_VIOLATIONS` is enabled, so if we catch one here, we can assume it's enabled without actually checking the context in this method.

I'd like to introduce an interface for exceptions that would simplify this maybe, but I'll wait until we're sure the current approach is correct.

> Or is the problem that these error messages are currently a bit "internal" (e.g. `Failed to denormalize attribute foo for class Foo\Bar`) and so the new exception message allows denormalizers to advertise a safer, "validation" message? Or is there another purpose?
>
> If the only purpose is to advertise a validation-safe message, perhaps we could add a new optional property+method (thinking out loud) to `NotNormalizableValueException` that contains this - e.g. `getValidationMessage()`? Or maybe I'm way off understanding the mechanics :)

Indeed, the purpose is to advertise validation-safe messages (and maybe more validation related data later). Not sure `NotNormalizableValueException` is suitable for that though: currently, denormalizers might throw other exceptions (e.g. `UnexpectedValueException`). We could either:
- always throw `NotNormalizableValueException`, but replacing a thrown exception with another would be a BC break I guess;
- add the methods you suggest to all possible exceptions.

WDYT?